### PR TITLE
Align erasure of `Array[Nothing]` and `Array[Null]` with Scala 2

### LINKED ISSUE #22515: Dotty Type Erasure for Scala 2 is wrong for `Array[Null]` and `Array[Nothing]`
## Compiler version

3.0.0-3.6.3

## Minimized code

```scala
//> using scala 3.6.3
@main def Test = Foo.a
```

```scala
//> using scala 2.13.16
object Foo {
    def a: Array[scala.Nothing] = ???
}
```

## Output

```scala
Exception in thread "main" java.lang.NoSuchMethodError: 'scala.runtime.Nothing$[] Foo$.a()'
	at playground$package$.Test(playground.scala:1)
	at Test.main(playground.scala:1)
```

## Expectation

Work at runtime. The issue was introduced by https://github.com/scala/scala3/pull/11846

## Notes
Same if we have in the return type `Array[scala.Null]`

### PR DESCRIPTION
**Pull request by hamzaremmal**
Closes #22515

### CODE REVIEW

**Review by smarter — APPROVED** [review]
Instead of doing a new scripted test, can you include the test code in https://github.com/scala/scala3/tree/main/sbt-test/scala2-compat/erasure ?

**Comment by smarter** [comment]
(also the test could include `Array[? <: Nothing]` and `Array[? <: Null]`)

File: sbt-test/scala2-compat/erasure/scala2Lib/Api.scala [hunk 1]
```diff
@@ -186,4 +186,10 @@ class Z {
   def objectARRAY_88(x: Array[Any]): Unit = {}
   def objectARRAY_89(x: Array[AnyRef]): Unit = {}
   def objectARRAY_90(x: Array[AnyVal]): Unit = {}
+
+  def objectARRAY_91(x: Array[Nothing]): Unit = {}
+  def objectARRAY_92(x: Array[Null]): Unit = {}
+  def objectARRAY_93(x: Array[_ <: Nothing]): Unit = {}
+  def objectARRAY_94(x: Array[_ <: Null]): Unit = {}
```
Comment (hamzaremmal) [inline sbt-test/scala2-compat/erasure/scala2Lib/Api.scala:193]: This should actually be:
```suggestion
  def nothing$ARRAY_93(x: Array[_ <: Nothing]): Unit = {}
  def null$ARRAY_94(x: Array[_ <: Null]): Unit = {}
```
but for some reason, it passes the test locally, I will have to test it better, but my machine keeps crashing... so checking with the CI :)

Comment (hamzaremmal) [reply sbt-test/scala2-compat/erasure/scala2Lib/Api.scala:193]: So this is weird, the CI passes while `Array[_ <: Nothing]` erasure is wrong. I've tested with the following:

```scala
// In Scala 2
class Foo {
    def a: Array[Nothing] = null
    def b: Array[Null] = null
    def c: Array[_ <: Nothing] = null
    def s: Array[_ <: Null] = null
}
```


```scala
// In Scala 3
@main def Test =
    (new Foo).a
    (new Foo).b
    (new Foo).c
    (new Foo).s
```

The generated bytecode from Scala 2 is the following:
```java
Compiled from "s2.scala"
public class Foo {
  public java.lang.Object[] a();
  public java.lang.Object[] b();
  public scala.runtime.Nothing$[] c();
  public scala.runtime.Null$[] s();
  public Foo();
}
```

We can clearly see Scala 3 (this PR) emitting a call to the wrong method for `Foo.c`:
```
public final class s3$package$ implements java.io.Serializable {
  public static final s3$package$ MODULE$;

  public static {};
    Code:
       0: new           #2                  // class s3$package$
       3: dup
       4: invokespecial #16                 // Method "<init>":()V
       7: putstatic     #18                 // Field MODULE$:Ls3$package$;
      10: return

  public scala.runtime.Null$[] Test();
    Code:
       0: new           #29                 // class Foo
       3: dup
       4: invokespecial #30                 // Method Foo."<init>":()V
       7: invokevirtual #34                 // Method Foo.a:()[Ljava/lang/Object;
      10: pop
      11: new           #29                 // class Foo
      14: dup
      15: invokespecial #30                 // Method Foo."<init>":()V
      18: invokevirtual #37                 // Method Foo.b:()[Ljava/lang/Object;
      21: pop
      22: new           #29                 // class Foo
      25: dup
      26: invokespecial #30                 // Method Foo."<init>":()V
      29: invokevirtual #40                 // Method Foo.c:()Ljava/lang/Object;
      32: pop
      33: new           #29                 // class Foo
      36: dup
      37: invokespecial #30                 // Method Foo."<init>":()V
      40: invokevirtual #43                 // Method Foo.s:()[Lscala/runtime/Null$;
      43: areturn
}
```

Comment (hamzaremmal) [reply sbt-test/scala2-compat/erasure/scala2Lib/Api.scala:193]: I've checked a bit and throwing a `LinkageError` will not be considered a failure... but the problem is that there is a linkage error before the code I've added (I still need to find which one).

Comment (hamzaremmal) [reply sbt-test/scala2-compat/erasure/scala2Lib/Api.scala:193]: There is already an issue opened on sbt (https://github.com/sbt/sbt/issues/6852). The proposed workaround (`fork := true`) worked locally

File: sbt-test/scala2-compat/erasure/dottyApp/Main.scala [hunk 2]
```diff
@@ -53,10 +53,10 @@ object Main {
     z.c_40(dummy)
     z.c_41(dummy)
     z.c_42(dummy)
-    z.b_43(dummy)
+    //z.b_43(dummy)
```
Comment (hamzaremmal) [inline sbt-test/scala2-compat/erasure/dottyApp/Main.scala:56]: TODO: Inspect why both these methods' erasures are wrong, or why were they considered correct.

File: sbt-test/scala2-compat/erasure/build.sbt [hunk 3]
```diff
@@ -1,3 +1,5 @@
+ThisBuild / fork := true
```
Comment (hamzaremmal) [inline sbt-test/scala2-compat/erasure/build.sbt:1]: Workaround until #22524 is resolved

**Comment by hamzaremmal** [comment]
I forgot it was approved... I have the reflex to enable auto-merge when I'm done with a PR (Sorry...)

**Comment by smarter** [comment]
No worries, seems fine to me as is.

**Comment by tgodzik** [comment]
I think this should require minor @hamzaremmal ? Just for the sake of backporting to LTS since we are at 3.7.0 anyway

**Comment by sjrd** [comment]
Not necessarily. It's a bug fix to be able to call certain Scala 2 methods that would previously throw `AbstractMethodError`s.

**Comment by hamzaremmal** [comment]
It-s a bug fix, it shouldn't require a minor release

**Comment by tgodzik** [comment]
That's what I thought, but wanted to be sure. Thanks!
